### PR TITLE
Added support for connection error retries

### PR DESCRIPTION
**Pull request by whiskeysierra**
## Motivation and Context

Introduces automatic retries for *pre-request* exceptions that usually happen while a connection is being established. See #450 for details.

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [x] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:

- [x] My change requires a change to the documentation.
- [x] I have updated the documentation accordingly.
- [x] I have added tests to cover my changes.

### CODE REVIEW

File: riptide-failsafe/src/test/java/org/zalando/riptide/failsafe/FailsafePluginRetriesTest.java [hunk 1]
```diff
@@ -16,14 +15,22 @@
...
+import org.zalando.riptide.Attributes;
 import org.zalando.riptide.Http;
+import org.zalando.riptide.Plugin;
+import org.zalando.riptide.RequestExecution;
 import org.zalando.riptide.httpclient.ApacheClientHttpRequestFactory;
+import org.zalando.riptide.idempotency.IdempotencyPredicate;
 
 import javax.annotation.Nullable;
 import java.io.IOException;
+import java.net.NoRouteToHostException;
```
Comment (whiskeysierra) [inline riptide-failsafe/src/test/java/org/zalando/riptide/failsafe/FailsafePluginRetriesTest.java]: Codacy Issue found: [Avoid unused imports such as 'java.net.NoRouteToHostException'](https://app.codacy.com/manual/whiskeysierra/riptide/pullRequest?prid=4799382)

File: riptide-faults/src/test/java/org/zalando/riptide/faults/RuleTest.java [hunk 2]
```diff
@@ -0,0 +1,44 @@
+package org.zalando.riptide.faults;
+
+import org.junit.jupiter.api.Test;
+
+import static org.junit.jupiter.api.Assertions.*;
```
Comment (whiskeysierra) [inline riptide-faults/src/test/java/org/zalando/riptide/faults/RuleTest.java]: Codacy Issue found: [Avoid unused imports such as 'org.junit.jupiter.api.Assertions'](https://app.codacy.com/manual/whiskeysierra/riptide/pullRequest?prid=4799382)

File: riptide-failsafe/src/test/java/org/zalando/riptide/failsafe/FailsafePluginRetriesTest.java [hunk 3]
```diff
@@ -16,14 +15,22 @@
 import org.springframework.http.HttpStatus;
 import org.springframework.http.client.ClientHttpResponse;
 import org.springframework.http.converter.json.MappingJackson2HttpMessageConverter;
+import org.zalando.riptide.Attributes;
```
Comment (whiskeysierra) [inline riptide-failsafe/src/test/java/org/zalando/riptide/failsafe/FailsafePluginRetriesTest.java]: Codacy Issue found: [Avoid unused imports such as 'org.zalando.riptide.Attributes'](https://app.codacy.com/manual/whiskeysierra/riptide/pullRequest?prid=4799382)

File: README.md [hunk 4]
```diff
@@ -275,7 +275,7 @@ Riptide comes with a way to register extensions in the form of plugins.
 - [`AuthorizationPlugin`](#riptide-auth), adds `Authorization` support
 - [`FailsafePlugin`](riptide-failsafe), adds retries, circuit breaker, backup requests and timeout support
 - [`MicrometerPlugin`](riptide-micrometer), adds metrics for request duration
-- [`TransientFaultPlugin`](riptide-faults), detects transient faults, e.g. network issues
+- [`TransientFaults`](riptide-faults), detects transient faults, e.g. network issues
```
Comment (lukasniemeier-zalando) [inline README.md]: Is it a `Plugin` or not?

**Comment by lukasniemeier-zalando** [comment]
:+1:

File: README.md (hunk 4 above)
Comment (whiskeysierra) [reply README.md]: Technically not, but I'm inclined to leave it there for now

**Pull request merged**